cloudsmith/data_source_package_download.go

```diff
@@ -0,0 +1,150 @@
+func getPackageURL(pc *providerConfig, namespace string, repository string, query string, packageName string, packageVersion string) (string, string, error) {
+	req := pc.APIClient.PackagesApi.PackagesList(pc.Auth, namespace, repository)
+
+	queryString := query
+
+	if packageVersion != "" {
+		queryWithVersion := fmt.Sprintf("%s version:%s", query, packageVersion)
+		queryString = queryWithVersion
+	}
+
+	req = req.Query(queryString)
+	req = req.PageSize(1)
+
+	packagesList, _, err := pc.APIClient.PackagesApi.PackagesListExecute(req)
+	if err != nil {
+		return "", "", err
+	}
+
+	for _, pkg := range packagesList {
+		if pkg.GetName() == packageName {
+			return pkg.GetCdnUrl(), pkg.GetFilename(), nil
+		}
+	}
+
+	return "", "", errors.New("package not found")
+}
+
+func dataSourcePackageDownloadRead(d *schema.ResourceData, m interface{}) error {
+	pc := m.(*providerConfig)
+
+	namespace := d.Get("namespace").(string)
+	repository := d.Get("repository").(string)
+	query := d.Get("query").(string)
+	packageName := d.Get("package_name").(string)
+	packageVersion := d.Get("package_version").(string)
+	destinationPath := d.Get("destination_path").(string)
+
+	cdnURL, filename, err := getPackageURL(pc, namespace, repository, query, packageName, packageVersion)
+	if err != nil {
+		return err
+	}
+
+	apiKey := pc.GetAPIKey()
+
+	// Use the filename variable when constructing the destination file path
+	destinationFilepath := filepath.Join(destinationPath, filename)
+
+	// Download the file to the destination path
+	err = downloadFile(destinationFilepath, cdnURL, apiKey)
+	if err != nil {
+		return err
+	}
+
+	d.SetId(fmt.Sprintf("%s/%s/%s", namespace, repository, packageName))
+	return nil
+}
+
+func downloadFile(filepath string, url string, apiKey string) error {
+	req, err := http.NewRequest(http.MethodGet, url, nil)
+	if err != nil {
+		return err
+	}
+
+	req.Header.Add("Authorization", fmt.Sprintf("Token %s", apiKey))
+
+	client := &http.Client{}
+	resp, err := client.Do(req)
+	if err != nil {
+		return err
+	}
+	defer resp.Body.Close()
+
+	if resp.StatusCode != http.StatusOK {
+		return fmt.Errorf("failed to download file: %s, status code: %d", url, resp.StatusCode)
+	}
+
+	out, err := os.Create(filepath)
+	if err != nil {
+		return err
+	}
+	defer out.Close()
+
+	_, err = io.Copy(out, resp.Body)
+	return err
+}
+
+func dataSourcePackageDownload() *schema.Resource {
+	return &schema.Resource{
+		Read: dataSourcePackageDownloadRead,
+
+		Schema: map[string]*schema.Schema{
+			"repository": {
+				Type:         schema.TypeString,
+				Description:  "The repository to which the package belongs.",
+				Required:     true,
+				ValidateFunc: validation.StringIsNotEmpty,
+			},
+			"namespace": {
+				Type:         schema.TypeString,
+				Description:  "The namespace to which the package belongs.",
+				Required:     true,
+				ValidateFunc: validation.StringIsNotEmpty,
+			},
```

Comment:
can we use organization rather than namesapce here? i know the provider is inconsistent, but so is the app itself, and we're slowing migrating away from `namespace` over time